zerver/lib/actions.py

```diff
@@ -3600,14 +3600,32 @@ def internal_send_huddle_message(
     return message_ids[0]
 
 
-def pick_color(used_colors: Set[str]) -> str:
-    # These colors are shared with the palette in stream_settings_ui.js.
+def pick_colors(
+    used_colors: Set[str], color_map: Dict[int, str], recipient_ids: List[int]
+) -> Dict[int, str]:
+    used_colors = set(used_colors)
+    recipient_ids = sorted(recipient_ids)
+    result = {}
+
+    other_recipient_ids = []
+    for recipient_id in recipient_ids:
+        if recipient_id in color_map:
+            color = color_map[recipient_id]
+            result[recipient_id] = color
+            used_colors.add(color)
+        else:
+            other_recipient_ids.append(recipient_id)
+
     available_colors = [s for s in STREAM_ASSIGNMENT_COLORS if s not in used_colors]
 
-    if available_colors:
-        return available_colors[0]
-    else:
-        return STREAM_ASSIGNMENT_COLORS[len(used_colors) % len(STREAM_ASSIGNMENT_COLORS)]
+    for i, recipient_id in enumerate(other_recipient_ids):
+        if i < len(available_colors):
+            color = available_colors[i]
+        else:
+            color = STREAM_ASSIGNMENT_COLORS[i % len(STREAM_ASSIGNMENT_COLORS)]
```

Comment:
I think this introduces a subtle bug: If you're adding a single stream to a user, we will now always use `STREAM_ASSIGNMENT_COLORS[0]`, rather than `num_streams % len(STREAM_ASSIGNMENT_COLORS)` (aka a different color each time), since for each call, `i=0`.

It may be hard to verify this in the UI, since I think colors are picked by the webapp, but should be testable by using the API to create colors (or using a test).

Reply:
Hmm, I'm not sure how this might have worked before -- we'd need to be using a list data structure, not a set data structure, to avoid that outcome.  So it seems possible this isn't a regression.

Reply:
This code definitely worked before we did bulk-add, but that was ~2014.  I don't remember exactly when we added sets, but the set logic would have clearly broken the modulo logic as soon as any user had more than 24 streams.  I think I explain pretty clearly in the commit message how this whole situation is mostly hypothetical and irrelevant.